File: wordle.c
#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <stdbool.h>

#define MAXCHARSPERLINE 30

char guess[5];


int loadGuess()
{
  printf("Enter 5 characters, using the following:\n");
  printf("  _ --> UNKNOWN\n");
  printf("  a --> known letter\n");
  printf("  A --> known letter and position\n");
  printf("\n");
  scanf("%5s", guess);
  if (strlen(guess) < 5)
  {
    return -1;
  }

  printf("\n\nYou entered: %s\n\n", guess);

  return 0;
}

bool wordContainsLetter(char word[5], char letter)
{
  for (int i=0; i<5; i++)
  {
    if (word[i] == letter) return true;
  }

  return false;
}

bool wordContainsLetterInPosition(char *word[5], char *letter, int position)
{
  return word[position] == tolower(*letter);
  for (int i=0; i<5; i++)
  {
    if (word[i] == letter) return true;
  }

  return false;
}

int solve()
{
  puts("Found these words:\n\n");
  
  FILE *fp;
  char word[MAXCHARSPERLINE];

  /* opening file for reading */
  fp = fopen("wordlist.txt", "r");
  if (fp == NULL)
  {
    perror("Error opening file");
    return (-1);
  }

  while (fgets(word, MAXCHARSPERLINE, fp) != NULL)
  {
    // wordle uses 5-letter words only -- include terminating char
    if (strlen(word) == 6)
    {
      int wordIsMatch = 1;  // assume it's a match until it isn't

      // iterate over guess and find matches
      for (int guessLetter=0; guessLetter < 5; guessLetter++)
      {
        char c = guess[guessLetter];

        if (c != '_')
        {
          // is guess lower or upper case?
          bool isUpper = (c >= 'A' && c <= 'Z');

          // iterate over word and see if it contains letter guess
          int wordIncludesGuessLetter = 0;

          for (int wordLetter=0; wordLetter < 5; wordLetter++)
          {
            // match letters found anywhere in the word except the stated position if lower case
            if (!isUpper) 
            {
              if (word[wordLetter] == c && wordLetter != guessLetter)
              {
                wordIncludesGuessLetter = 1;
                break;  // break out of word iterator
              }
            }
            else
            {
              if (word[wordLetter] == c && wordLetter == guessLetter)
              {
                wordIncludesGuessLetter = 1;
                break;  // break out of word iterator
              }
            }
          }

          if (wordIncludesGuessLetter == 0) 
          {
            wordIsMatch = 0;
            break;  // break out of guess iterator
          }
        }
      }
      
      if (wordIsMatch == 1)
        puts(word);
    }
  }

  fclose(fp);

  return 0;
}


int main(int argc, char *argv[])
{
  int ok;
  ok = loadGuess();
  if (ok == -1)
  {
    perror("Invalid guess. You need to enter a 5-character guess.\n");
    return -1;
  }

  ok = solve();
  return ok;
}

File: wordle.js
const fs = require('fs');
const { exit } = require('process');
const dictFilespec = 'dictionary.txt';

// get guess
const guess = process.argv[2];
const eliminated = process.argv[3];

if (!guess || guess.length != 5) {
  console.log(`USAGE:
  node wordle.js GUESS
    GUESS = '_', 'a', 'A'
      _ is unknown
      a is known
      A is known position
  
  First guesses:
    SLATE
    CRANE
    TRACE
    SLANT
    CRATE
    CARTE

  Second guesses:
    DEMON
    MOLDY
    MONTH
    MODEL
    MELON
  `);
  exit(1);
}

// get 5 letter words with no eliminated letters
const wordList = (fs.readFileSync(dictFilespec, {encoding: 'utf-8'}))
  .split('\n')
  .filter(w => w.length == 5)
  .filter(w => {
    for (let e = 0; e < eliminated.length; e++) {
      if (w.toLowerCase().includes(eliminated[e].toLowerCase())) return false;
    }

    return true;
  })

const wordHasLetterInPosition = (word, letter, position) => {
  return word[position].toLowerCase() == letter.toLowerCase();
}

// search for possible matches
for (let word of wordList) {
  word = word.toLowerCase();
  let wordIsMatch = true;

  for (let i = 0; i < 5; i++) {
    const letter = guess[i];

    if (letter == '_') {
      continue;
    }

    else if (letter.toLowerCase() == letter && word.includes(letter) && word[i] != letter) {      
      continue;
    }

    else if (letter.toUpperCase() == letter && word[i] == letter.toLowerCase()) {
      continue;
    }

    wordIsMatch = false;
  }

  if (wordIsMatch) console.log(word);
}